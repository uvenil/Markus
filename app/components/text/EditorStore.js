// @flow
'use strict';

import { extendObservable } from 'mobx';
import Record from '../../data/Record';
import Config from '../../definitions/config.json';
import Rx from 'rx-lite';

export default class EditorStore {
    _changes            : Rx.Subject;
    editorShown         : boolean;
    previewShown        : boolean;
    highlightActiveLine : boolean;
    tabSize             : number;
    useSoftTabs         : boolean;
    wordWrap            : boolean;
    showLineNumbers     : boolean;
    showPrintMargin     : boolean;
    printMarginColumn   : number;
    showInvisibles      : boolean;
    showFoldWidgets     : boolean;
    showGutter          : boolean;
    scrollPastEnd       : boolean;
    record              : ?Record;
    theme               : string;
    fontFamily          : ?string;
    textSize            : number|string;
    cursorPosition      : string;
    isOverwriteEnabled  : boolean;

    constructor() {
        this._changes = new Rx.Subject();

        this.highlightActiveLine = Config.defaultHighlightActiveLine;
        this.tabSize             = Config.defaultTabSize;
        this.useSoftTabs         = Config.defaultUseSoftTabs;
        this.wordWrap            = Config.defaultWordWrap;
        this.showLineNumbers     = Config.defaultShowLineNumbers;
        this.showPrintMargin     = Config.defaultShowPrintMargin;
        this.printMarginColumn   = Config.defaultPrintMarginColumn;
        this.showInvisibles      = Config.defaultShowInvisibles;
        this.showFoldWidgets     = Config.defaultShowFoldWidgets;
        this.showGutter          = Config.defaultShowGutter;
        this.scrollPastEnd       = Config.defaultScrollPastEnd;

        extendObservable(this, {
            editorShown        : true,
            previewShown       : true,
            record             : undefined,
            theme              : Config.defaultTheme,
            fontFamily         : undefined,
            textSize           : undefined,
            cursorPosition     : undefined,
            isOverwriteEnabled : false
        });
    }

    get changes() : any {
        return this._changes;
    }
}
