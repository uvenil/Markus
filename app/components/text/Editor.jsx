// @flow
'use strict';

import React from 'react';
import { observer } from 'mobx-react';
import muiThemeable from 'material-ui/styles/muiThemeable';
import AceEditor from 'react-ace';
import ReactMarkdown from 'react-markdown';
import CodeRenderer from './renderers/CodeRenderer.jsx';
import EditorStore from './EditorStore';
import Record from '../../data/Record';
import EventUtils from '../../utils/EventUtils';
import Constants from '../../Constants';
import { showTextBoxContextMenu } from '../../utils/ContextMenuUtils';
import merge from 'lodash.merge';

require('brace/ext/searchbox');
require('brace/ext/spellcheck');
require('brace/ext/whitespace');

@observer
class Editor extends React.Component {
    static propTypes : Object;

    _events         : any;
    _handleChange   : Function;
    _init           : Function;
    _changeSettings : Function;
    _changeFont     : Function;
    _handleRefresh  : Function;

    constructor(props : Object) {
        super(props);

        this._events = [];

        this._handleChange = (value : string) => {
            if (this.props.store.record) {
                this.props.store.record.update(value);
            } else {
                this.props.store.record = Record.fromText(value);
            }

            this.props.store.changes.onNext(this.props.store.record);
        };

        this._init = (data : Object) => {
            this.refs.editor.editor.setHighlightActiveLine(data.highlightActiveLine);
            this.refs.editor.editor.$updateHighlightActiveLine();
            this.refs.editor.editor.getSession().setTabSize(data.tabSize);
            this.refs.editor.editor.getSession().setUseSoftTabs(data.useSoftTabs);
            this.refs.editor.editor.getSession().setUseWrapMode(data.wordWrap);
            this.refs.editor.editor.renderer.showLineNumbers = data.showLineNumebrs;
            this.refs.editor.editor.renderer.$gutterLayer.setShowLineNumbers(data.showLineNumebrs);
            this.refs.editor.editor.setShowPrintMargin(data.showPrintMargin);
            this.refs.editor.editor.setPrintMarginColumn(data.printMarginColumn);
            this.refs.editor.editor.renderer.$loop.schedule(this.refs.editor.editor.renderer.CHANGE_GUTTER);
            this.refs.editor.editor.setShowInvisibles(data.showInvisibles);
            this.refs.editor.editor.setShowFoldWidgets(data.showFoldWidgets);
            this.refs.editor.editor.renderer.setShowGutter(data.showGutter);
            this.refs.editor.editor.renderer.$scrollPastEnd = data.scrollPastEnd;

            this.refs.editor.editor.session.selection.on('changeCursor', () : void => this.props.store.cursorPosition = this.refs.editor.editor.getCursorPosition());
            this.refs.editor.editor.session.on('changeOverwrite', () : void => this.props.store.isOverwriteEnabled = this.refs.editor.editor.getOverwrite());

            this.refs.editor.editor.container.addEventListener('contextmenu', event => {
                event.preventDefault();

                showTextBoxContextMenu();

                return false;
            });

            this.refs.editor.editor.container.style.lineHeight = Constants.EDITOR_LINE_HEIGHT;
            this.refs.editor.editor.renderer.updateFontSize();
        };

        this._changeSettings = (data : Object) => {
            if (data.name === 'highlightActiveLine') {
                this.refs.editor.editor.setHighlightActiveLine(data.value);
                this.refs.editor.editor.$updateHighlightActiveLine();
            } else if (data.name === 'tabSize') {
                this.refs.editor.editor.getSession().setTabSize(data.value);
            } else if (data.name === 'useSoftTabs') {
                this.refs.editor.editor.getSession().setUseSoftTabs(data.value);
            } else if (data.name === 'wordWrap') {
                this.refs.editor.editor.getSession().setUseWrapMode(data.value);
            } else if (data.name === 'showLineNumbers') {
                this.refs.editor.editor.renderer.showLineNumbers = data.value;
                this.refs.editor.editor.renderer.$gutterLayer.setShowLineNumbers(data.value);
                this.refs.editor.editor.renderer.$loop.schedule(this.refs.editor.editor.renderer.CHANGE_GUTTER);
            } else if (data.name === 'showPrintMargin') {
                this.refs.editor.editor.setShowPrintMargin(data.value);
            } else if (data.name === 'printMarginColumn') {
                this.refs.editor.editor.setPrintMarginColumn(data.value);
            } else if (data.name === 'showInvisibles') {
                this.refs.editor.editor.setShowInvisibles(data.value);
            } else if (data.name === 'showFoldWidgets') {
                this.refs.editor.editor.setShowFoldWidgets(data.value);
            } else if (data.name === 'showGutter') {
                this.refs.editor.editor.renderer.setShowGutter(data.value);
            } else if (data.name === 'scrollPastEnd') {
                this.refs.editor.editor.renderer.$scrollPastEnd = data.value;
            } else {
                console.warn('Unrecognized setting ' + data.name + ' = ' + data.value);
            }
        };

        this._changeFont = (font : string) => this.refs.editor.editor.setOptions({ fontFamily : font });

        this._handleRefresh = () : void => this.refs.editor.editor.resize();
    }

    componentDidMount() : void {
        this._events.push(EventUtils.register('editor.init', data => this._init(data)));
        this._events.push(EventUtils.register('editor.settings.change', data => this._changeSettings(data)));
        this._events.push(EventUtils.register('editor.font.change', font => this._changeFont(font)));
        this._events.push(EventUtils.register('editor.refresh', () => this._handleRefresh()));
    }

    componentWillUnmount() : void {
        this._events.forEach(event => EventUtils.unregister(event));
    }

    render() : any {
        require('brace/mode/markdown');
        require('brace/theme/' + this.props.store.theme);

        const style = {
            width  : '100%',
            height : 'calc(100vh - ' + (Constants.BOTTOM_BAR_HEIGHT + 1) + 'px)'
        };

        return (
            <div style={merge(style, this.props.style)}>
                <div
                    style={{
                        display : this.props.store.record ? 'table' : 'none',
                        width   : '100%'
                    }}>
                    {/* Editor */}
                    <AceEditor
                        ref="editor"
                        mode="markdown"
                        theme={this.props.store.theme}
                        value={this.props.store.record ? this.props.store.record.fullText : undefined}
                        width={this.props.store.editorShown ? this.props.store.previewShown ? '50%' : '100%' : '0'}
                        height={'calc(100vh - ' + (Constants.TOP_BAR_HEIGHT + Constants.BOTTOM_BAR_HEIGHT * (this.props.store.record ? 3 : 1)) + 'px)'}
                        fontSize={this.props.store.textSize}
                        showGutter={this.props.store.showGutter}
                        highlightActiveLine={this.props.store.highlightActiveLine}
                        showPrintMargin={this.props.store.showPrintMargin}
                        tabSize={this.props.tabSize}
                        editorProps={{
                            fontFamily          : this.props.store.fontFamily,
                            $blockScrolling     : 'Infinity',
                            $showLineNumbers    : this.props.store.showLineNumbers,
                            $printMarginColumn  : this.props.store.printMarginColumn,
                            $showInvisibles     : this.props.store.showInvisibles,
                            $scrollPastEnd      : this.props.store.scrollPastEnd,
                            $useSoftTabs        : this.props.store.useSoftTabs,
                            $wrap               : this.props.store.wordWrap
                        }}
                        style={{
                            display          : this.props.store.editorShown ? 'table-cell' : 'none',
                            borderRightWidth : 1,
                            borderRightStyle : 'solid',
                            borderRightColor : this.props.muiTheme.palette.borderColor
                        }}
                        onChange={value => this._handleChange(value)} />
                    {/* Preview */}
                    <div
                        style={{
                            display      : this.props.store.previewShown ? 'table-cell' : 'none',
                            width        : this.props.store.previewShown ? this.props.store.editorShown ? '50%' : '100%' : '0',
                            paddingLeft  : Constants.PADDING_X1,
                            paddingRight : Constants.PADDING_X1,
                            color        : this.props.muiTheme.palette.textColor
                        }}>
                        <ReactMarkdown
                            source={this.props.store.record ? this.props.store.record.fullText : undefined}
                            renderers={{
                                Code      : CodeRenderer,
                                CodeBlock : CodeRenderer
                            }} />
                    </div>
                </div>
                <div
                    style={{
                        display         : this.props.store.record ? 'none' : 'block',
                        width           : '100%',
                        height          : 'calc(100vh - ' + (Constants.BOTTOM_BAR_HEIGHT + 1) + 'px)',
                        backgroundColor : this.props.muiTheme.palette.primary3Color
                    }} />
            </div>
        );
    }
}

Editor.propTypes = {
    store : React.PropTypes.instanceOf(EditorStore).isRequired,
    style : React.PropTypes.object
};

export default muiThemeable()(Editor);
