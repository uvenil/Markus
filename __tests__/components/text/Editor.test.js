'use strict';

import React from 'react';
import Renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import lightBaseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import Editor from '../../../app/components/text/Editor.jsx';
import EditorStore from '../../../app/components/text/EditorStore';

jest.mock('../../../app/utils/ContextMenuUtils');

it('renders correctly', () => {
    const store = new EditorStore();

    const component = Renderer.create(
        <MuiThemeProvider muiTheme={getMuiTheme(lightBaseTheme)}>
            <Editor store={store} />
        </MuiThemeProvider>
    ).toJSON();

    expect(component).toMatchSnapshot();
});
