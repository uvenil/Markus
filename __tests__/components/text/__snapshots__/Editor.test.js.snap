// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "height": "calc(100vh - 29px)",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "display": "none",
        "width": "100%",
      }
    }
  >
    <div
      id="brace-editor"
      style={
        Object {
          "borderRightColor": "#e0e0e0",
          "borderRightStyle": "solid",
          "borderRightWidth": 1,
          "display": "table-cell",
          "height": "calc(100vh - 66px)",
          "width": "50%",
        }
      }
    />
    <div
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "display": "table-cell",
          "paddingLeft": 8,
          "paddingRight": 8,
          "width": "50%",
        }
      }
    >
      <div />
    </div>
  </div>
  <div
    style={
      Object {
        "backgroundColor": "#bdbdbd",
        "display": "block",
        "height": "calc(100vh - 29px)",
        "width": "100%",
      }
    }
  />
</div>
`;
